packages/network/src/logic/proxy/ProxyStreamConnectionServer.ts

```diff
@@ -0,0 +1,109 @@
+import { StreamPartManager } from '../StreamPartManager'
+import { Event as NodeToNodeEvent, NodeToNode } from '../../protocol/NodeToNode'
+import { NodeId } from '../../identifiers'
+import { Event, Node } from '../Node'
+import {
+    ProxyConnectionRequest,
+    ProxyDirection,
+    StreamPartID,
+    UnsubscribeRequest
+} from '@streamr/protocol'
+import { Logger } from "@streamr/utils"
+import { Propagation } from '../propagation/Propagation'
+
+const logger = new Logger(module)
+
+export interface ProxyStreamConnectionServerOptions {
+    streamPartManager: StreamPartManager
+    nodeToNode: NodeToNode
+    propagation: Propagation
+    node: Node
+    acceptProxyConnections: boolean
+}
+
+interface ProxyConnection {
+    direction: ProxyDirection // Direction is from the clients point of view
+    userId: string
+}
+
+export class ProxyStreamConnectionServer {
+    private readonly streamPartManager: StreamPartManager
+    private readonly nodeToNode: NodeToNode
+    private readonly node: Node
+    private readonly acceptProxyConnections: boolean
+    private readonly connections: Map<StreamPartID, Map<NodeId, ProxyConnection>>
+    private readonly propagation: Propagation
+
+    constructor(opts: ProxyStreamConnectionServerOptions) {
+        this.streamPartManager = opts.streamPartManager
+        this.nodeToNode = opts.nodeToNode
+        this.node = opts.node
+        this.acceptProxyConnections = opts.acceptProxyConnections
+        this.propagation = opts.propagation
+        this.connections = new Map()
+        this.nodeToNode.on(NodeToNodeEvent.PROXY_CONNECTION_REQUEST_RECEIVED, (message,  nodeId) => {
+            this.processHandshakeRequest(message, nodeId)
+        })
+
+        this.nodeToNode.on(NodeToNodeEvent.LEAVE_REQUEST_RECEIVED, (message, nodeId) => {
+            this.processLeaveRequest(message, nodeId)
+        })
+    }
+
+    private async processHandshakeRequest(message: ProxyConnectionRequest, nodeId: NodeId): Promise<void> {
+        const streamPartId = message.getStreamPartID()
+        const isAccepted = this.streamPartManager.isSetUp(streamPartId) && this.acceptProxyConnections
```

Comment:
If `this.streamPartManager.isSetUp(streamPartId)` true because of some side effect of the fact that we accept proxy connections to that stream part? Or should we just check this.acceptProxyConnections here and setup the `streamPart` to `streamPartManager` if necessary?

Reply:
The server should not accept any proxy connections to streams that it has not already joined. This is one of the core requirements of the feature.